src/libponyrt/mem/pagemap.c

```diff
@@ -146,14 +208,28 @@ void ponyint_pagemap_set(const void* addr, chunk_t* chunk)
 #endif
     }
 
-    uintptr_t ix = ((uintptr_t)addr >> level[i].shift) & level[i].mask;
+    ix = ((uintptr_t)addr >> level[i].shift) & level[i].mask;
+    pony_assert(ix <= (uintptr_t)level[i].size);
+    pony_assert(ix <= (uintptr_t)level[i].mask);
+
+    // at the lowest level we have to double the index since we store
+    // two pointers (`chunk_t*` and `pony_actor_t*`) instead of one
+    if (i == PAGEMAP_LEVELS - 1)
+      ix = ix * 2;
     next_node = &(((PONY_ATOMIC_RVALUE(void*)*)node)[ix]);
   }
 
+  pony_assert(ix <= (uintptr_t)level[PAGEMAP_LEVELS - 1].size);
+  pony_assert(ix/2 <= (uintptr_t)level[PAGEMAP_LEVELS - 1].mask);
   atomic_store_explicit(next_node, chunk, memory_order_release);
+  ix++;
+  pony_assert(ix <= (uintptr_t)level[PAGEMAP_LEVELS - 1].size);
+  pony_assert(ix/2 <= (uintptr_t)level[PAGEMAP_LEVELS - 1].mask);
+  next_node = &(((PONY_ATOMIC_RVALUE(void*)*)node)[ix]);
+  atomic_store_explicit(next_node, actor, memory_order_release);
```

Comment:
note from sync: please document your understanding of the atomics and their safety here as you stated in the sync call.

Reply:
added a comment.. let me know if it's not clear enough..